feat(api): Adds `NoFilterCacheListMixin` for caching unfiltered list of results

ERosendo opened the pull request:
Introduces `NoFilterCacheListMixin` to improve performance of list endpoints by caching responses when there's **no pagination** and either **a count** is specifically requested **or** **no filters** are applied to the queryset. This prevents repeated database queries for frequently accessed unfiltered data, reducing load and speeding up response times.

The mixin first determines if caching is applicable, checking for the absence of pagination and the presence of either a count request or no active filters. If eligible, it attempts to retrieve the response from db_cache; a hit returns the cached data immediately. Otherwise, the generated response is added to the cache via a `post-render` callback.

This new mixin class will cache requests that meet the following criteria:

- Requests with no filters:
   - `https://www.courtlistener.com/api/rest/v4/dockets/`

- v4 count requests:
   
   - `http://courtlistener.com/api/rest/v4/dockets/?count=on`
   
- Requests with order_by and no other filters:
   
   - `http://courtlistener.com/api/rest/v4/dockets/?order_by=id`

The mixin will not cache the following requests:

- Requests with filters:
  
   - `http://courtlistener.com/api/rest/v4/dockets/?court=ca3`

- Requests with pagination:
  
  - `http://courtlistener.com/api/rest/v4/dockets/?cursor=cD02OTcyNjU3MQ%3D%3D`
  
fixes #5673

mlissner commented on cl/api/tests.py:
```diff
@@ -3456,3 +3463,158 @@ def test_anonymous_user_handling(self, mock_get_redis):
+@patch("cl.api.utils.get_cache_key_for_no_filter_mixin")
+class CacheListApiResponseTest(TestCase):
+    def setUp(self):
+        self.cache = caches["db_cache"]
```
Why not use redis cache during tests for speed?

ERosendo replied:
oh that's interesting, I'm just setting up a variable to avoid repetition in the tests. The cache connection within the mixin is not configurable and is fixed to the database cache.

mlissner commented on cl/api/utils.py:
```diff
@@ -450,6 +454,115 @@ def list(self, *args, **kwargs):
         return super().list(*args, **kwargs)
 
 
+def get_cache_key_for_no_filter_mixin(
```
I think I'd do:

```suggestion
def make_cache_key_for_no_filter_mixin(
```

ERosendo replied:
alright!

mlissner commented on cl/api/utils.py:
```diff
@@ -450,6 +454,115 @@ def list(self, *args, **kwargs):
+                response_obj (Response): The DRF Response object to be cached.
+            """
+            # Cache the response for 1 hour
+            cache_connection.set(key, response, 60 * 60)
```
I think we can do ten minutes here:

```suggestion
            cache_connection.set(key, response, 10 * 60)
```

ERosendo replied:
Sure! I'll update the timeout.

mlissner reviewed (approved):
Looks pretty good to me. Just a few comments.

We should bump the API version to 4.2 and document this, and we should apply this to all our endpoints, I think.

ERosendo commented:
@mlissner Thanks for your feedback! I've addressed your comments and added the new mixin to the `/opinions` API endpoint.

> We should bump the API version to 4.2 and document this

Let's create an issue to address this.

mlissner commented:
Thinking about it some more, I think we are fine not adding documentation and a release about this. It's a very narrow change, and doesn't impact the API in a way that'd affect users.
